kernel/thread.c

```diff
@@ -518,8 +518,13 @@ void z_setup_new_thread(struct k_thread *new_thread,
 			  prio, options);
 
 #ifdef CONFIG_SMP
-	/* switch handle must be non-null except when inside z_swap() */
-	new_thread->switch_handle = new_thread;
+	/* switch_handle must be non-null except when inside z_swap().
+	 * Not all architectures initialize the field, so do it here
+	 * as a fallback.
+	 */
+	if (new_thread->switch_handle == NULL) {
```

Comment:
Not a huge fan of covering mistakes in arch code with extra checking in the kernel. Couldn't we just fix the arch code that isn't doing this properly?

> Add a test, and clarify that this is a fallback and that really this
is the arch layer's job. 

I don't see a test case added? Or did you mean simply this if block?

Reply:
Also is CONFIG_SMP the right Kconfig to enclose this, or should this be inside an #ifdef CONFIG_SWITCH?

Reply:
Yes, I meant "test for the situation where the field is still null".

As far as fixing the arch layer: I agree, but unfortunately we can't build and run all the SMP architectures without external tooling, so there's some risk to patching stuff that is untested.

And the latter is sorta "both".  Obviously we're initializing a field that is controlled by CONFIG_USE_SWITCH, but the specific rule being enforced (that it must be non-null) is something that only SMP has code to detect.  And even then only SMP when CONFIG_MP_NUM_CPUS > 1, now that I think about it.

Reply:
> As far as fixing the arch layer: I agree, but unfortunately we can't build and run all the SMP architectures without external tooling, so there's some risk to patching stuff that is untested.

Which are we missing? ARC? We can have @vonhust assist with this, I think.